### ISSUE TITLE
Distributed training mode: the master node does training work but shouldn't

### ISSUE THREAD

**Issue by dakoner** [tensorflow/models#4382]
I spent some time debugging distributed training mode and noticed something.  I'm not sure if it's intentional or not.  In particular, when you start distributed training by setting a TF_CONFIG, the master node ends up evaluating training data, computing loss, and optimizing is.  This is in addition to the worker nodes.  The master node appears to do dual duty (master coordination in addition to training).

I've worked in distributed computing for a while, and the convention is that the master node doesn't do training work- it's harder to configure the resource requests for a master that does dual duty, and there are subtle performance problems as the training and the master coordination tend to fight for limited CPU availability.  It also plays poorly with frameworks that handle distributed training (for example, kubeflow has a TFJob that makes it easy to set up distributed training on k8s; it assumes the "master" container doesn't need or want GPU.

Before going further, I'd like to ask: is the choice of doing training work on the master node intentional?  I am guessing it's just an oversight, and that the code for converting the TF_CONFIG to a computation needs to explicitly ensure that work is only assigned to worker nodes.

If the choice is intentional, then I'll have to make adjustments to my distributed training pipeline to include additional resources for the master node (more CPU, along with GPU), and spend time debugging performance problems.  If the choice is not intentional, then i think we can proceed to repair the script.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by dakoner** [comment]
None of those values matter at all, they are just N/A.

Also, there isn't any issue template that I can see.

**Comment by qlzh727** [comment]
Thanks for submitting the issue. From https://www.tensorflow.org/deploy/distributed, it seems that the master node is more like the coordinator instead of real worker that executes the workload. The real workload should be executed by worker service. Can you provide the cluster setting here so that we can understand your config setting?

Also adding Derek here who is more knowledgeable about distribution strategy.

**Comment by mrry** [comment]
Indeed, it does seem that `object_detection/train.py` is hard-coded to treat the `"master"` job as another worker role. Reassigning to @derekjchow, who shows up as the original author of that code.

**Comment by ChChwang** [comment]
there seems should be a master in every worker and the re-using the master as worker. After a period of time, it raise error "Session was not ready after *** seconds". Why? what should I do ? Thanks a lot.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
